src/types/UiaTextRangeBase.cpp

```diff
@@ -310,14 +310,262 @@ IFACEMETHODIMP UiaTextRangeBase::ExpandToEnclosingUnit(_In_ TextUnit unit) noexc
+    {
+        LOG_HR(wil::ResultFromCaughtException());
+        UiaTracing::TextRange::FindAttribute(*this, attributeId, val, searchBackwards, static_cast<UiaTextRangeBase&>(**ppRetVal), UiaTracing::AttributeType::Error);
+        return E_INVALIDARG;
+    }
+
+    // Get some useful variables
+    const auto& buffer{ _pData->GetTextBuffer() };
+    const auto bufferSize{ buffer.GetSize() };
+    const auto inclusiveEnd{ _getInclusiveEnd() };
+
+    // Start/End for the resulting range.
+    // NOTE: we store these as "first" and "second" anchor because,
+    //       we just want to know what the inclusive range is.
+    //       We'll do some post-processing to fix this on the way out.
+    std::optional<COORD> resultFirstAnchor;
+    std::optional<COORD> resultSecondAnchor;
+
+    // Start/End for the direction to perform the search in
+    const auto searchStart{ searchBackwards ? inclusiveEnd : _start };
+    const auto searchEnd{ searchBackwards ? _start : inclusiveEnd };
+
+    // Iterate from searchStart to searchEnd in the buffer.
+    // If we find the attribute we're looking for, we update resultFirstAnchor/SecondAnchor appropriately.
+    const Viewport viewportRange{ _blockRange ? Viewport::FromDimensions(_start, inclusiveEnd.X - _start.X + 1, inclusiveEnd.Y - _start.Y + 1) : bufferSize };
+    auto iter{ buffer.GetCellDataAt(searchStart, viewportRange, searchEnd) };
+    for (; iter; searchBackwards ? --iter : ++iter)
+    {
+        const auto& attr{ iter->TextAttr() };
+        if (checkIfAttrFound(attr))
+        {
+            // populate the first anchor if it's not populated.
+            // otherwise, populate the second anchor.
+            if (!resultFirstAnchor.has_value())
+            {
+                resultFirstAnchor = iter.Pos();
+            }
+            else
+            {
+                resultSecondAnchor = iter.Pos();
+            }
+        }
+        else if (resultFirstAnchor.has_value() && resultSecondAnchor.has_value())
+        {
+            // Exit the loop early if...
+            // - the cell we're looking at doesn't have the attr we're looking for
+            // - the anchors have been populated
+            // This means that we've found a contiguous range where the text attribute was found.
+            // No point in searching through the rest of the search space.
+            break;
+        }
+    }
+
+    // If a result was found, populate ppRetVal with the UiaTextRange
+    // representing the found selection anchors.
+    if (resultFirstAnchor.has_value() && resultSecondAnchor.has_value())
+    {
+        RETURN_IF_FAILED(Clone(ppRetVal));
+        UiaTextRangeBase& range = static_cast<UiaTextRangeBase&>(**ppRetVal);
+
+        // IMPORTANT: resultFirstAnchor and resultSecondAnchor make up an inclusive range.
+        range._start = searchBackwards ? *resultSecondAnchor : *resultFirstAnchor;
+        range._end = searchBackwards ? *resultFirstAnchor : *resultSecondAnchor;

```

Comment:
Yeah `std::swap` is a bit slower, as Carlos' ternary solution is entirely branchless (code like that doesn't actually use branches/jumps, but rather `cmove`).
But if it's not called more than X times per second, instruction-level performance like that certainly doesn't matter. 😄

Reply:
> entirely branchless

On x86-64, right? No guarantee about arm64/x86?

Reply:
`cmov` exists since Pentium II actually! On a Zen3 this has a 1 cycle latency (how long it takes that you can read the result) and 0.5 cycles for throughput (how long it takes until you can issue another instruction of the same kind), which makes `cmov` basically almost as fast as `mov` itself!

I don't know what AArch64 uses - I believe there are no conditional moves at all...
But since ARM designs are generally a bit more effective I doubt that the lack of `cmov` hurts it too much.

Reply:
Ah I just found out that ARM64/AArch64 actually introduced `csel`, or "conditional select"! It has just the same performance benefits. As expected MSVC compiles the ternary operator to `csel`, but fails to do so for `std::swap`.